replica/table.cc

```diff
@@ -2325,9 +2325,14 @@ future<> table::clear() {
 // NOTE: does not need to be futurized, but might eventually, depending on
 // if we implement notifications, whatnot.
 future<db::replay_position> table::discard_sstables(db_clock::time_point truncated_at) {
-    assert(std::ranges::all_of(compaction_groups(), [this] (const compaction_group& cg) {
-        return _compaction_manager.compaction_disabled(cg.as_table_state());
-    }));
+    // truncate_table_on_all_shards() disables compaction for the truncated
+    // tables and views, so we normally expect compaction to be disabled on
+    // this table. But as shown in issue #17543, it is possible that a new
+    // materialized view was created right after truncation started, and it
+    // would not have compaction disabled when this function is called on it.
+    //assert(std::ranges::all_of(compaction_groups(), [this] (const compaction_group& cg) {
+    //    return _compaction_manager.compaction_disabled(cg.as_table_state());
+    //}));
```

Comment:
Shouldn't we at least warn? Or check if it's a view?

Reply:
I can change the assert to on_internal_error, and leave it if the table is not a view. It still feels very wrong to me, but I guess it may catch some regression in the future that has nothing to do with views. Ok.

Reply:
I'm not sure I see the value - what will happen when we bump into on_internal_error()? It'll crash on QA and in production throw an exception?

Reply:
Yes

Reply:
Pushed a new version with on_internal_error() if not a view, ignored when a view.

If on_internal_error() fails in production, the user will see a failure (exception) in the "TRUNCATE" operation, which, arguably, exactly what we expect to happen. I'm not sure, though, what happens if we have this failure on just one node and not others - I think we can end up with a half-truncated state, but don't think there's a better solution - a crash is worse.